src/loader.rs

```diff
@@ -6,61 +6,68 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
+//! Defines an `AnalysisLoader` trait, which allows to specify directories
+//! from which save-analysis JSON files can be read. Also supplies a 
+//! default implementation `CargoAnalysisLoader` for Cargo-emitted save-analysis
+//! files.
+
 use std::ffi::OsStr;
+use std::fmt;
 use std::path::{Path, PathBuf};
 use std::process::Command;
-use std::sync::Mutex;
 
-use raw::Target;
 use AnalysisHost;
 
 #[derive(Debug)]
 pub struct CargoAnalysisLoader {
-    pub path_prefix: Mutex<Option<PathBuf>>,
+    pub path_prefix: Option<PathBuf>,
     pub target: Target,
 }
 
+impl CargoAnalysisLoader {
+    pub fn new(target: Target) -> CargoAnalysisLoader {
+        CargoAnalysisLoader {
+            path_prefix: None,
+            target,
+        }
+    }
+}
+
+/// Allows to specify from where and which analysis files will be considered
+/// when reloading data to lower.
 pub trait AnalysisLoader: Sized {
     fn needs_hard_reload(&self, path_prefix: &Path) -> bool;
     fn fresh_host(&self) -> AnalysisHost<Self>;
-    fn set_path_prefix(&self, path_prefix: &Path);
+    fn set_path_prefix(&mut self, path_prefix: &Path);
     fn abs_path_prefix(&self) -> Option<PathBuf>;
-    fn iter_paths<F, T>(&self, f: F) -> Vec<T>
-    where
-        F: Fn(&Path) -> Vec<T>;
+    /// Returns every directory in which paths are to be considered.
+    /// TODO: Can it also return files?
+    fn paths(&self) -> Vec<PathBuf>;
```

Comment:
Could `paths` be renamed `search_directories` or something a bit more descriptive?

Reply:
Sure, I wasn't sure before if it's designed to only return directories or can it also return paths, so that's the most neutral name I could come up with